data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,381 @@
+        )
+        local_path = df_to_tempfile(df)
+        return local_path
+
+
+def get_response(
+    user_agent: str | None,
+    type_of_data: TYPE_OF_DATA,
+    records_from: str,
+    max_records: str,
+    response_columns: list[str] | None,
+    search_criteria: list[CriteriaValue | CriteriaRange],
+    num_retries: int,
+) -> pd.DataFrame | None:
+    """
+    Makes a request call to api and returns data in pandas dataframe.
+    """
+    request_dict = create_request_dict(
+        type_of_data, records_from, max_records, response_columns, search_criteria
+    )
+    xml = dict_to_xml_obj(request_dict)
+    xml_string = xml_obj_to_string(xml)
+    headers = {"Content-Type": "application/xml"}
+    if user_agent:
+        headers["User-Agent"] = user_agent
+
+    for n in range(num_retries):
+        try:
+            response = requests.post(url=api_endpoint, data=xml_string, headers=headers)
+            response.raise_for_status()
+            break
+        except requests.exceptions.RequestException as e:
+            response = None
+            user_agent = random.choice(USER_AGENTS)
+            print(
+                f"Failed to get a response from api (try # {n + 1}/{num_retries}). Exception: {e}"
+            )
+            # pausisng before next request: we don't want to overload API
+            time.sleep(random.randint(10, 20))
+
+    if response:
+        xml_response = response.text
+        try:
+            df = pd.read_xml(
+                xml_response, xpath="./result_records/spending_transactions/*"
+            )
+            return df
+        except Exception as e:
+            print(
+                f"Exception happened when reading xml into pandas. Xml body: {xml_response}"
+            )
+            return None
+    else:
+        return None
+
+
+def create_request_dict(
+    type_of_data: TYPE_OF_DATA,
+    records_from: str,
+    max_records: str,
+    response_columns: list[str] | None,
+    search_criteria: list[CriteriaValue | CriteriaRange],
+) -> dict:
+    """
+    Creates a request body in dictionary format.
+
+    Sample output dict: {
+        'request': {
+            'type_of_data':'Spending',
+            'records_from': '1',
+            'response_columns': [
+                {'column': 'agency'},
+                {'column': 'expense_category'}
+            ]
+        }
+    }
+    """
+    request_dict: dict[str, Any] = {
+        "request": {
+            "type_of_data": type_of_data,
+            "records_from": records_from,
+            "max_records": max_records,
+        }
+    }
+    if response_columns:
+        request_dict["request"]["response_columns"] = [
+            {"column": col} for col in response_columns
+        ]
+    if search_criteria:
+        request_dict["request"]["search_criteria"] = [
+            {"criteria": criteria_item} for criteria_item in search_criteria
+        ]
+
+    return request_dict
+
+
+def dict_to_xml_obj(request_dict: dict) -> ET.Element:
```

Comment:
Curious as to Alex and Damon's thoughts on both, my thoughts are
1. This would be good to have in dcpy, but for the moment the dcpy - data-library connection seems a little prickly, so I wouldn't include that in this PR. But I think something along the lines of `dcpy.utils.xml` would make sense longer term
2. In general my thought on docstrings is more thorough is never a bad thing. Main potential drawback is if the code is changed, you have to remember to update the example, where mypy won't catch errors for you. But this seems like a function that wouldn't change much, if it was say one of our connectors I would be a bit more hesitant just because we've been moving pretty quickly and I wouldn't be surprised if they fell out of sync by accident at some point